cli/util/upload_state.go

```diff
@@ -0,0 +1,209 @@
+
+	params := pool.DefaultOptions()
+	params.SetHealthcheckTimeout(neofs.DefaultHealthcheckTimeout)
+	params.SetNodeDialTimeout(neofs.DefaultDialTimeout)
+	params.SetNodeStreamTimeout(neofs.DefaultStreamTimeout)
+	p, err := pool.New(pool.NewFlatNodeParams(rpcNeoFS), signer, params)
+	if err != nil {
+		return cli.Exit(fmt.Sprintf("failed to create NeoFS pool: %v", err), 1)
+	}
+	pWrapper := poolWrapper{p}
+	if err = pWrapper.Dial(context.Background()); err != nil {
+		return cli.Exit(fmt.Sprintf("failed to dial NeoFS pool: %v", err), 1)
+	}
+	defer p.Close()
+
+	var containerObj container.Container
+	err = retry(func() error {
+		containerObj, err = p.ContainerGet(ctx.Context, containerID, client.PrmContainerGet{})
+		return err
+	}, maxRetries, debug)
+	if err != nil {
+		return cli.Exit(fmt.Errorf("failed to get container with ID %s: %w", containerID, err), 1)
+	}
+	containerMagic := containerObj.Attribute("Magic")
+
+	logger := zap.NewExample()
+	cfg, err := options.GetConfigFromContext(ctx)
+	if err != nil {
+		return cli.Exit(err, 1)
+	}
+
+	chain, store, err := server.InitBlockChain(cfg, logger)
+	if err != nil {
+		return cli.Exit(err, 1)
+	}
+
+	magic := strconv.Itoa(int(chain.GetConfig().Magic))
+	if containerMagic != magic {
+		return cli.Exit(fmt.Sprintf("container magic %s does not match the network magic %s", containerMagic, magic), 1)
+	}
+	syncInterval := uint32(chain.GetConfig().ProtocolConfiguration.StateSyncInterval)
+	stateObjCount, err := searchStateIndex(ctx, pWrapper, containerID, acc.PrivateKey(), attr, uint(syncInterval), maxRetries, debug)
+
+	stateModule := chain.GetStateModule()
+	currentHeight := stateModule.CurrentLocalHeight()
+
+	currentStateIndex := currentHeight / syncInterval
+	if currentStateIndex <= stateObjCount {
+		fmt.Fprintf(ctx.App.Writer, "No new states to upload. Last state sync at block %d, current height %d\n",
+			stateObjCount*syncInterval, currentHeight)
+		return nil
+	}
+
+	fmt.Fprintf(ctx.App.Writer, "Latest state object found at height %d, current height %d, %d new states to upload\n", stateObjCount*syncInterval, currentHeight, currentStateIndex-stateObjCount)
+
+	for state := stateObjCount; state < currentStateIndex; state++ {
+		stateRoot, err := stateModule.GetStateRoot(state * syncInterval)
+		if err != nil {
+			return cli.Exit(fmt.Sprintf("failed to get state root at height %d: %v", state, err), 1)
+		}
+
+		nodes, err := traverseRawMPT(stateRoot.Root, store, mpt.ModeLatest)
+		if err != nil {
+			return cli.Exit(fmt.Sprintf("failed to traverse MPT at height %d: %v", state*syncInterval, err), 1)
+		}
+
+		objBytes, err := encodeMPTNodes(nodes)
+		if err != nil {
+			return cli.Exit(fmt.Sprintf("failed to encode MPT nodes at height %d: %v", state*syncInterval, err), 1)
+		}
+
+		h, err := chain.GetHeader(chain.GetHeaderHash(state * syncInterval))
+		if err != nil {
+			return cli.Exit(fmt.Sprintf("failed to get header at height %d: %v", state*syncInterval, err), 1)
+		}
```

Comment:
Move it to the start of the command, fail-fast, because MPT traversal is the heaviest part.